relay-cardinality/src/redis.rs

```diff
@@ -341,6 +418,250 @@ impl SlidingWindow {
+        if window != inner.current_window {
+            return CacheRead::Empty;
+        }
+
+        CacheRead::Read {
+            inner,
+            organization_id,
+            limit,
+        }
+    }
+
+    /// Acquires a write lock from the cache and returns an update handle.
+    ///
+    /// All operations done on the handle share the same lock. To release the lock
+    /// the returned [`CacheUpdate`] must be dropped.
+    fn update<'a>(
+        &'a self,
+        organization_id: OrganizationId,
+        item_scope: &'a str,
+        window: u64,
+    ) -> CacheUpdate<'a> {
+        let mut inner = self.inner.write().unwrap_or_else(PoisonError::into_inner);
+
+        // If the passed window is newer then the current window, reset the cache to the new window.
+        if window > inner.current_window {
+            inner.current_window = window;
+            inner.cache.clear();
+        } else if window != inner.current_window {
+            // If the window is older don't do anything and give up the lock early.
+            return CacheUpdate::Empty;
+        }
+
+        CacheUpdate::Update {
+            inner,
+            key: CacheKeyRef {
+                organization_id,
+                item_scope,
+            },
+        }
+    }
+}
+
+/// Cache read handle.
+///
+/// Holds a cache read lock, the lock is released on drop.
+enum CacheRead<'a> {
+    Empty,
+    Read {
+        inner: RwLockReadGuard<'a, CacheInner>,
+        organization_id: OrganizationId,
+        limit: usize,
+    },
+}
+
+impl<'a> CacheRead<'a> {
+    fn check(&self, item_scope: &str, hash: u32) -> CacheOutcome {
+        match self {
+            CacheRead::Empty => CacheOutcome::Unknown,
+            CacheRead::Read {
+                inner,
+                organization_id,
+                limit,
+            } => inner
+                .cache
+                .get(&CacheKeyRef {
+                    organization_id: *organization_id,
+                    item_scope,
+                })
+                .map_or(CacheOutcome::Unknown, |s| s.check(hash, *limit)),
+        }
+    }
+}
+
+/// Cache update handle.
+///
+/// Holds a cache write lock, the lock is released on drop.
+enum CacheUpdate<'a> {
+    Empty,
+    Update {
+        inner: RwLockWriteGuard<'a, CacheInner>,
+        key: CacheKeyRef<'a>,
+    },
+}
+
+impl<'a> CacheUpdate<'a> {
+    /// Marks a hash as accepted in the cache, future checks of the item will immediately mark the
+    /// item as accepted.
+    pub fn accept(&mut self, hash: u32) {
+        if let Self::Update { inner, key } = self {
+            inner.cache.entry_ref(key).or_default().0.insert(hash);
+        }
+    }
+}
+
+/// Critical section of the [`Cache`].
+#[derive(Debug, Default)]
+struct CacheInner {
+    cache: hashbrown::HashMap<CacheKey, CacheValue>,
+    current_window: u64,
+}
+
+/// Scope specific information of the cache.
+#[derive(Debug, Default)]
+struct CacheValue(hashbrown::HashSet<u32>);
```

Comment:
I was confused by the term `Value` here for a moment. Given that this _is_ the cache for a given scope, should we name it `ScopedCache` or something like that?